src/codegen/x86_64/arch_x86_64.c

```diff
@@ -880,11 +898,70 @@ static void emit_instruction(CodegenContext *context, IRInstruction *inst) {
   case IR_NOT: {
     femit_reg(context, I_NOT, inst->operand->result);
     enum RegSize size = regsize_from_bytes(type_sizeof(inst->operand->type));
-    femit_reg_to_reg(context, I_MOV, inst->operand->result, inst->result, size);
+    femit_reg_to_reg(context, I_MOV, inst->operand->result, size, inst->result, size);
+  } break;
+  case IR_ZERO_EXTEND: {
+    usz operand_byte_size = type_sizeof(inst->operand->type);
+    usz result_byte_size = type_sizeof(inst->type);
+    ASSERT (result_byte_size > operand_byte_size, "Zero extension result must be larger than operand");
+
+    // TODO: Handle things larger than a register, somehow... may need
+    // to push/pop registers...
```

Comment:
If we’re *zero-extending* to something that doesn’t fit in a register, then there’s something wrong with the frontend, candidly. Sext/zext/trunc only really makes sense w/ pointers and integers, i.e. operands that will always fit in a register. If it’s trying to *zero-extend* a struct to an `integer` or sth, then something went wrong somewhere candidly...

Reply:
Integers may be larger than a register, in case of i128/u128, right? We don't have those yet, but no reason to disallow them :eyes:. I agree struct type sign extension makes no sense, but having integer types larger than the biggest register doesn't seem that uncommon.

Reply:
I think it might be best to treat those as structs internally or sth. Either way, adding support for `i128` seems like it will be a massive pain that I’d rather not think about rn honestly...

Reply:
:eyes: but then we need to be able to sign/zero extend structs for that to work, right?

Also, why would integer types larger than what fits in a register be a huge deal?

Reply:
Btw, these comments were only here because I copied `IR_COPY`, IIRC 😛

Reply:
At some point we just start calling out to GNU MP and call it good lmao. That makes a lot of sense, thank you